x-pack/plugins/security_solution/public/timelines/store/middlewares/timeline_save.ts

```diff
@@ -0,0 +1,297 @@
+      } else if (key === 'columns' && get(key, timeline) != null) {
+        return set(
+          key,
+          get(key, timeline).map((col: ColumnHeaderOptions) =>
+            omit(['initialWidth', 'width', '__typename', 'esTypes'], col)
+          ),
+          acc
+        );
+      } else if (key === 'filters' && get(key, timeline) != null) {
+        const filters = get(key, timeline);
+        return set(
+          key,
+          filters != null
+            ? filters.map((myFilter: Filter) => {
+                const basicFilter = omit(['$state'], myFilter);
+                return {
+                  ...basicFilter,
+                  meta: {
+                    ...basicFilter.meta,
+                    field:
+                      (isMatchAllFilter(basicFilter) ||
+                        isPhraseFilter(basicFilter) ||
+                        isPhrasesFilter(basicFilter) ||
+                        isRangeFilter(basicFilter)) &&
+                      basicFilter.meta.field != null
+                        ? convertToString(basicFilter.meta.field)
+                        : null,
+                    value:
+                      basicFilter.meta.value != null
+                        ? convertToString(basicFilter.meta.value)
+                        : null,
+                    params:
+                      basicFilter.meta.params != null
+                        ? convertToString(basicFilter.meta.params)
+                        : null,
+                  },
+                  ...(isMatchAllFilter(basicFilter)
+                    ? {
+                        query: {
+                          match_all: convertToString(
+                            (basicFilter as MatchAllFilter).query.match_all
+                          ),
+                        },
+                      }
+                    : { match_all: null }),
+                  ...(isExistsFilter(basicFilter) && basicFilter.query.exists != null
+                    ? { query: { exists: convertToString(basicFilter.query.exists) } }
+                    : { exists: null }),
+                  ...((isQueryStringFilter(basicFilter) || get('query', basicFilter) != null) &&
+                  basicFilter.query != null
+                    ? { query: convertToString(basicFilter.query) }
+                    : { query: null }),
+                  ...(isRangeFilter(basicFilter) && basicFilter.query.range != null
+                    ? { query: { range: convertToString(basicFilter.query.range) } }
+                    : { range: null }),
+                  ...(isScriptedRangeFilter(basicFilter) &&
+                  basicFilter.query.script !=
+                    null /* TODO remove it when PR50713 is merged || esFilters.isPhraseFilter(basicFilter) */
+                    ? { query: { script: convertToString(basicFilter.query.script) } }
+                    : { script: null }),
```

Comment:
any way we could improve this? it is very hard to read

Reply:
I wish...I'm a little afraid to touch this one since I didn't write it and I'm not sure its tests cover all corner cases.

Reply:
Yea, this is a nightmare. Each nested ternary should either be it's own function or at the very least have a description of what's going on. I wonder if there are better utilities for doing this now too since this code is relatively outdated... 🤔

Reply:
maybe it would be a good idea to add a lot of tests THEN clean it up?

Reply:
agree with that, in addition I would like to point out 2 things: 1) this code is pretty similar to the insight code that serializes markdown to filters, so much so that we should absolutely combine the two paths in someway in the future 2) there's a pretty clear bug to me here, in that we are not handling the `isCombinedFilter` case at all here, added to support nested filters, and so there is no way we are saving and restoring timelines with those types of filters correctly, afaik not a known issue right now.

Reply:
actually it does work, since we are copying the entire $meta part of the filter over, and then just using whatever was stored in the saved object for combined filter types. not sure how much of this code is needed at all really then.

Reply:
I think we could consider all of this outside of the scope of this PR. Like @janmonschke mentioned above, this code wasn't really modified, just moved around.
As much as I'd like to see this changes, it could be part of a broader effort: how to we want our timeline store to look like...

Reply:
Yes agree completely it's out of scope, I just wanted to point those 2 things out, especially the not handling isCombinedFilter case.